splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java

```diff
@@ -29,108 +29,128 @@
-		TimestampOracle to = _instance;
-		if(to==null){
-			synchronized(TimestampOracle.class){
-				to = _instance;
-				if(to==null){
-					SpliceLogUtils.info(LOG,"Initializing TimestampOracle...");
-					to=_instance=new TimestampOracle(timestampBlockManager,blockSize);
-				}
-			}
-		}
-		return to;
-	}
-	
-	private TimestampOracle(TimestampBlockManager timestampBlockManager, int blockSize) throws TimestampIOException {
+        TimestampOracle to = _instance;
+        if(to==null){
+            synchronized(TimestampOracle.class){
+                to = _instance;
+                if(to==null){
+                    SpliceLogUtils.info(LOG,"Initializing TimestampOracle...");
+                    to=_instance=new TimestampOracle(timestampBlockManager,blockSize);
+                }
+            }
+        }
+        return to;
+    }
+
+    private TimestampOracle(TimestampBlockManager timestampBlockManager, int blockSize) throws TimestampIOException {
         this.timestampBlockManager=timestampBlockManager;
         this.blockSize = blockSize * TIMESTAMP_INCREMENT;
-		initialize();
-	}
+        initialize();
+    }
 
     /**
      * Read the current state of the block from the timestampBlockManager
      *
      * @throws TimestampIOException
      */
-	private void initialize() throws TimestampIOException {
-			synchronized(this) {
+    private void initialize() throws TimestampIOException {
+            synchronized(this) {
                 _maxReservedTimestamp = timestampBlockManager.initialize();
-				_timestampCounter.set(_maxReservedTimestamp + TIMESTAMP_INCREMENT);
-			}
-			try {
-				registerJMX();
-			} catch (Exception e) {
+                _timestampCounter.updateAndGet(oldTS -> Math.max(oldTS, _maxReservedTimestamp + TIMESTAMP_INCREMENT));
+            }
+            try {
+                registerJMX();
+            } catch (Exception e) {
                 SpliceLogUtils.error(LOG, "Unable to register Timestamp Generator with JMX. Service will function but metrics will not be available.");
-			}
+            }
+    }
+
+    public long getCurrentTimestamp() {
+        return _timestampCounter.get();
+    }
+
+    private static long roundUp(long timestamp) {
+        assert timestamp >= 0;
+        long remainder = timestamp % TIMESTAMP_INCREMENT;
+        if (remainder == 0) {
+            return timestamp;
+        }
+        return timestamp - remainder + TIMESTAMP_INCREMENT;
+    }
+
+    public synchronized void bumpTimestamp(long newTimestamp) throws TimestampIOException {
+        newTimestamp = roundUp(newTimestamp);
+        if (newTimestamp > _timestampCounter.get()) {
+            if (newTimestamp> _maxReservedTimestamp) {
+                reserveNextBlock(newTimestamp);
+            }
+            long currentTimestampCounter;
+            while ((currentTimestampCounter = _timestampCounter.get()) < newTimestamp) {
+                _timestampCounter.compareAndSet(currentTimestampCounter, newTimestamp);
```

Comment:
_compareAndSet_ returns _true_ if successful, which also guarantees the condition is satisfied.
Consider:

	public synchronized void bumpTimestamp(long newTimestamp) throws TimestampIOException {
		newTimestamp = roundUp(newTimestamp);
		if (newTimestamp> _maxReservedTimestamp) {
			reserveNextBlock(newTimestamp);
		}
		long currentTimestampCounter;
		while ((currentTimestampCounter = _timestampCounter.get()) < newTimestamp) {
			if (_timestampCounter.compareAndSet(currentTimestampCounter, newTimestamp)) break;
		}
	}

Reply:
done